### ISSUE TITLE
multiplication and softmax throws incorrect results under torch.compile

### ISSUE THREAD

**Issue by maybeLee** [pytorch/pytorch#161150]
### 🐛 Describe the bug

Maybe this issue is caused by the optimization related to FMA.

```python
import torch
torch.manual_seed(100)


class M(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.scale = torch.tensor([-1.7e14])

    def forward(self, x):
        for _ in range(3):
            x = x * self.scale
        softmaxed = torch.softmax(x, dim=0)
        return x, softmaxed


m = M()
opt = torch.compile(m)
inp = torch.randn((1,10))

m_out = m(inp)
opt_out = opt(inp)

print(f"[Eager] Softmax Input: ", m_out[0])
print(f"[Eager] Softmax output: ", m_out[-1])

print(f"[Compile] Softmax Input: ", opt_out[0])
print(f"[Compile] Softmax output: ", opt_out[-1])
```

```
[Eager] Softmax Input:  tensor([[-inf, inf, inf, -inf, inf, inf, inf, inf, -inf, inf]])
[Eager] Softmax output:  tensor([[nan, nan, nan, nan, nan, nan, nan, nan, nan, nan]])
[Compile] Softmax Input:  tensor([[-inf, inf, inf, -inf, inf, inf, inf, inf, -inf, inf]])
[Compile] Softmax output:  tensor([[1., 1., 1., 1., 1., 1., 1., 1., 1., 1.]])
```

Even though the softmax's input is not normal, the compiled model's softmax output is normal but an incorrect result.

### Versions

[pip3] numpy==2.0.2
[pip3] nvidia-cublas-cu12==12.8.4.1
[pip3] nvidia-cuda-cupti-cu12==12.8.90
[pip3] nvidia-cuda-nvrtc-cu12==12.8.93
[pip3] nvidia-cuda-runtime-cu12==12.8.90
[pip3] nvidia-cudnn-cu12==9.10.2.21
[pip3] nvidia-cufft-cu12==11.3.3.83
[pip3] nvidia-curand-cu12==10.3.9.90
[pip3] nvidia-cusolver-cu12==11.7.3.90
[pip3] nvidia-cusparse-cu12==12.5.8.93
[pip3] nvidia-cusparselt-cu12==0.7.1
[pip3] nvidia-nccl-cu12==2.27.3
[pip3] nvidia-nvjitlink-cu12==12.8.93
[pip3] nvidia-nvtx-cu12==12.8.90
[pip3] torch==2.8.0
[pip3] triton==3.4.0
[conda] numpy                     2.0.2                    pypi_0    pypi
[conda] nvidia-cublas-cu12        12.8.4.1                 pypi_0    pypi
[conda] nvidia-cuda-cupti-cu12    12.8.90                  pypi_0    pypi
[conda] nvidia-cuda-nvrtc-cu12    12.8.93                  pypi_0    pypi
[conda] nvidia-cuda-runtime-cu12  12.8.90                  pypi_0    pypi
[conda] nvidia-cudnn-cu12         9.10.2.21                pypi_0    pypi
[conda] nvidia-cufft-cu12         11.3.3.83                pypi_0    pypi
[conda] nvidia-curand-cu12        10.3.9.90                pypi_0    pypi
[conda] nvidia-cusolver-cu12      11.7.3.90                pypi_0    pypi
[conda] nvidia-cusparse-cu12      12.5.8.93                pypi_0    pypi
[conda] nvidia-cusparselt-cu12    0.7.1                    pypi_0    pypi
[conda] nvidia-nccl-cu12          2.27.3                   pypi_0    pypi
[conda] nvidia-nvjitlink-cu12     12.8.93                  pypi_0    pypi
[conda] nvidia-nvtx-cu12          12.8.90                  pypi_0    pypi
[conda] torch                     2.8.0                    pypi_0    pypi
[conda] triton                    3.4.0                    pypi_0    pypi

**Comment by dsashidh** [comment]
I'm working on this

**Comment by dsashidh** [comment]
Submitting PR soon

**Comment by isuruf** [comment]
This is an unfortunate result of trying to prevent overflow in `torch/_inductor/fx_passes/joint_graph.py`.

See https://github.com/pytorch/pytorch/blob/8abc2af9b963d72d56a3ba8ea455511b5676519d/torch/_inductor/fx_passes/joint_graph.py#L820-L835

**Comment by karthickai** [comment]
**🤖 PT2 Repro Agent**

I tried to reproduce this issue on PyTorch nightly (`2.13.0.dev20260516+cu132`) on NVIDIA H100 and confirmed it reproduces (3/3 runs failed).

All 3 runs reproduced the reported numerical mismatch. Eager softmax output is [nan, nan, nan, nan, nan, nan, nan, nan, nan, nan] (10 NaNs) while compiled softmax output is [1., 1., 1., 1., 1., 1., 1., 1., 1., 1.] (0 NaNs). Both compute the same intermediate `x * scale` tensor of +/-inf values, but their softmax outputs disagree on every element and the NaN masks are not equal. Re-running with TORCHINDUCTOR_EMULATE_PRECISION_CASTS=1 produced the same divergence (exit code 1).

The compiler bisector identified the offending component:
- **Backend**: `inductor`
- **Subsystem**: `joint_graph_passes`
- **Detail**: `pass_pattern_1`

<details>
<summary>Per-subsystem coverage (2 tested)</summary>

| Subsystem | Bug fixed when disabled? |
|---|---|
| `pre_grad_passes` | ❌ no |
| `joint_graph_passes` | ✅ yes |

Replay: `TORCH_BISECT_BACKEND=inductor TORCH_BISECT_SUBSYSTEM=joint_graph_passes TORCH_BISECT_MAX=7 python repro.py`

_Tool: `torch._inductor.compiler_bisector`_
</details>

<details>
<summary>Error output</summary>

```
[Eager] Softmax Input:   tensor([[-inf, inf, inf, -inf, inf, inf, inf, inf, -inf, inf]])
[Eager] Softmax output:  tensor([[nan, nan, nan, nan, nan, nan, nan, nan, nan, nan]])
[Compile] Softmax Input:   tensor([[-inf, inf, inf, -inf, inf, inf, inf, inf, -inf, inf]])
[Compile] Softmax output:  tensor([[1., 1., 1., 1., 1., 1., 1., 1., 1., 1.]])
Eager nan count:    10
Compiled nan count: 0
NaN masks equal:    False

ERROR conda.cli.main_run:execute(49): `conda run python /home/karthickps/.repro_agent/workdir/pytorch_pytorch/161150/repro.py` failed. (See above for error)
```
</details>

<details>
<summary>Repro script I used</summary>

```python
import os
import torch


class M(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.scale = torch.tensor([-1.7e14])

    def forward(self, x):
        for _ in range(3):
            x = x * self.scale
        softmaxed = torch.softmax(x, dim=0)
        return x, softmaxed


def test_fn() -> bool:
    """Return True if NO bug (eager == compiled). False if bug present."""
    torch._dynamo.reset()
    backend = os.environ.get("TORCH_COMPILE_BACKEND", "inductor")
    torch.manual_seed(100)
    m = M()
    opt = torch.compile(m, backend=backend)
    inp = torch.randn((1, 10))
    m_out = m(inp)
    opt_out = opt(inp)
    eager_softmax = m_out[-1]
    compiled_softmax = opt_out[-1]
    nan_masks_equal = torch.equal(
        torch.isnan(eager_softmax), torch.isnan(compiled_softmax)
    )
    if not nan_masks_equal:
        return False
    if not torch.allclose(
        eager_softmax, compiled_softmax, atol=1e-4, equal_nan=True
    ):
        return False
    return True


if __name__ == "__main__":
    torch.manual_seed(100)
    m = M()
    opt = torch.compile(m)
    inp = torch.randn((1, 10))
    m_out = m(inp)
    opt_out = opt(inp)
    print("[Eager] Softmax Input:  ", m_out[0])
    print("[Eager] Softmax output: ", m_out[-1])
    print("[Compile] Softmax Input:  ", opt_out[0])
    print("[Compile] Softmax output: ", opt_out[-1])
    eager_softmax = m_out[-1]
    compiled_softmax = opt_out[-1]
    eager_nan_mask = torch.isnan(eager_softmax)
    compiled_nan_mask = torch.isnan(compiled_softmax)
    print(f"Eager nan count:    {int(eager_nan_mask.sum())}")
    print(f"Compiled nan count: {int(compiled_nan_mask.sum())}")
    print(f"NaN masks equal:    {torch.equal(eager_nan_mask, compiled_nan_mask)}")
    try:
        ok = test_fn()
    except Exception:
        import traceback
        traceback.print_exc()
        os._exit(2)
    os._exit(0 if ok else 1)

```
</details>

<details>
<summary>My environment</summary>

```
Collecting environment information...
PyTorch version: 2.13.0.dev20260516+cu132
Is debug build: False
CUDA used to build PyTorch: 13.2
ROCM used to build PyTorch: N/A

OS: CentOS Stream 9 (x86_64)
GCC version: (GCC) 11.5.0 20240719 (Red Hat 11.5.0-14)
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.34

Python version: 3.13.12 | packaged by Anaconda, Inc. | (main, Feb 24 2026, 16:13:31) [GCC 14.3.0] (64-bit runtime)
Python platform: Linux-6.16.1-0_fbk2_hardened_0_gf40efc324cc8-x86_64-with-glibc2.34
Is CUDA available: True
CUDA runtime version: 12.8.93
CUDA_MODULE_LOADING set to: 
GPU models and configuration: GPU 0: NVIDIA H100
Nvidia driver version: 580.126.09
cuDNN version: Could not collect
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           52 bits physical, 57 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  46
On-line CPU(s) list:                     0-45
Vendor ID:                               AuthenticAMD
Model name:                              AMD EPYC 9654 96-Core Processor
CPU family:                              25
Model:                                   17
Thread(s) per core:                      1
Core(s) per socket:                      46
Socket(s):                               1
Stepping:                                1
BogoMIPS:                                4792.79
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm rep_good nopl cpuid extd_apicid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm cmp_legacy svm cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw perfctr_core ssbd ibrs ibpb stibp ibrs_enhanced vmmcall fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves avx512_bf16 clzero xsaveerptr wbnoinvd arat npt lbrv nrip_save tsc_scale vmcb_clean flushbyasid pausefilter pfthreshold v_vmsave_vmload vgif vnmi avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq la57 rdpid fsrm flush_l1d arch_capabilities
Virtualization:                          AMD-V
Hypervisor vendor:                       KVM
Virtualization type:                     full
L1d cache:                               2.9 MiB (46 instances)
L1i cache:                               2.9 MiB (46 instances)
L2 cache:                                23 MiB (46 instances)
L3 cache:                                736 MiB (46 instances)
NUMA node(s):                            1
NUMA node0 CPU(s):                       0-45
Vulnerability Gather data sampling:      Not affected
Vulnerability Ghostwrite:                Not affected
Vulnerability Indirect target selection: Not affected
Vulnerability Itlb multihit:             Not affected
Vulnerability L1tf:                      Not affected
Vulnerability Mds:                       Not affected
Vulnerability Meltdown:                  Not affected
Vulnerability Mmio stale data:           Not affected
Vulnerability Old microcode:             Not affected
Vulnerability Reg file data sampling:    Not affected
Vulnerability Retbleed:                  Not affected
Vulnerability Spec rstack overflow:      Vulnerable
Vulnerability Spec store bypass:         Vulnerable
Vulnerability Spectre v1:                Vulnerable: __user pointer sanitization and usercopy barriers only; no swapgs barriers
Vulnerability Spectre v2:                Vulnerable; IBPB: disabled; STIBP: disabled; PBRSB-eIBRS: Not affected; BHI: Not affected
Vulnerability Srbds:                     Not affected
Vulnerability Tsa:                       Vulnerable
Vulnerability Tsx async abort:           Not affected

Versions of relevant libraries:
[pip3] mypy_extensions==1.1.0
[pip3] numpy==2.4.4
[pip3] nvidia-cublas==13.4.0.1
[pip3] nvidia-cuda-cupti==13.2.75
[pip3] nvidia-cuda-nvrtc==13.2.78
[pip3] nvidia-cuda-runtime==13.2.75
[pip3] nvidia-cudnn-cu13==9.20.0.48
[pip3] nvidia-cufft==12.2.0.46
[pip3] nvidia-curand==10.4.2.55
[pip3] nvidia-cusolver==12.2.0.1
[pip3] nvidia-cusparse==12.7.10.1
[pip3] nvidia-cusparselt-cu13==0.8.1
[pip3] nvidia-nccl-cu13==2.29.7
[pip3] nvidia-nvjitlink==13.2.78
[pip3] nvidia-nvtx==13.2.75
[pip3] torch==2.13.0.dev20260516+cu132
[pip3] torchao==0.17.0
[pip3] torchaudio==2.11.0.dev20260516+cu132
[pip3] torchdata==0.11.0
[pip3] torchmetrics==1.0.3
[pip3] torchrec==1.7.0.dev20260413+cu130
[pip3] torchtune==0.6.1
[pip3] torchvision==0.28.0.dev20260516+cu132
[pip3] triton==3.7.0+git88b227e2
[conda] Could not collect
```
</details>

---
If the repro agent produced an incorrect result (wrong classification, bad repro script, missed the actual bug, etc.), add the label `repro:agent:h100:incorrect` to the issue.
